File: Sources/Xorswift/xorswift.swift

import Cxorswift

/// Generate single random UInt32 number.
public func xorshift() -> UInt32 {
    var ret: UInt32 = 0
    xorshift(&ret, 1)
    return ret
}

/// Generate random UInt32 numbers.
/// - Precondition: 
///   - `count` >= 0
public func xorshift(start: UnsafeMutablePointer<UInt32>, count: Int) {
    precondition(count >= 0, "Invalid argument: `count` must not be less than 0.")

    xorshift(start, Int32(count))
}

/// Sample random Float number from unifrom distribution [low, high).
/// - Precondition:
///   - `low` < `high`
public func xorshift_uniform(low: Float = 0,
                             high: Float = 1) -> Float {
    precondition(low < high, "Invalid argument: must be `low` < `high`")
    
    var ret: Float = 0
    xorshift_uniform(&ret, 1, low, high)
    return ret
}

/// Sample random Float numbers from unifrom distribution [low, high).
/// - Precondition: 
///   - `count` >= 0
///   - `low` < `high`
public func xorshift_uniform(start: UnsafeMutablePointer<Float>,
                             count: Int,
                             low: Float = 0,
                             high: Float = 1) {
    precondition(low < high, "Invalid argument: must be `low` < `high`")
    precondition(count >= 0, "Invalid argument: `count` must not be less than 0.")
    
    xorshift_uniform(start, Int32(count), low, high)
}

#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
    import Accelerate
    
    /// Sample random numbers from normal distribution N(mu, sigma).
    /// - Precondition:
    ///   - `count` >= 0
    ///   - `sigma` >= 0
    public func xorshift_normal(start: UnsafeMutablePointer<Float>,
                                count: Int,
                                mu: Float = 0,
                                sigma: Float = 1) {
        
        precondition(sigma >= 0, "Invalid argument: `sigma` must not be less than 0.")
        precondition(count >= 0, "Invalid argument: `count` must not be less than 0.")
        
        var _count = Int32(count)
        let __count = vDSP_Length(count)
        
        let buf1 = UnsafeMutablePointer<UInt32>.allocate(capacity: count)
        defer { buf1.deallocate(capacity: count) }
        xorshift(buf1, _count)
        vDSP_vfltu32(buf1, 1, start, 1, __count)
        
        let buf2 = UnsafeMutablePointer<Float>.allocate(capacity: count)
        defer { buf2.deallocate(capacity: count) }
        xorshift(buf1, _count)
        vDSP_vfltu32(buf1, 1, buf2, 1, __count)
        
        // X, Y in (0, 1)
        var divisor: Float = nextafter(Float(UInt32.max), Float(UInt64.max))
        var flt_min = Float.leastNonzeroMagnitude
        vDSP_vsdiv(start, 1, &divisor, start, 1, __count)
        vDSP_vsadd(start, 1, &flt_min, start, 1, __count)
        vDSP_vsdiv(buf2, 1, &divisor, buf2, 1, __count)
        vDSP_vsadd(buf2, 1, &flt_min, buf2, 1, __count)
        
        // sigma*sqrt(-2*log(X))
        vvlogf(start, start, &_count)
        var minus2sigma2 = -2 * sigma * sigma
        vDSP_vsmul(start, 1, &minus2sigma2, start, 1, __count)
        vvsqrtf(start, start, &_count)
        
        // cos(2*pi*Y)
        var pi2 = 2 * Float.pi
        vDSP_vsmul(buf2, 1, &pi2, buf2, 1, __count)
        vvcosf(buf2, buf2, &_count)
        
        vDSP_vmul(start, 1, buf2, 1, start, 1, __count)
        
        if(mu != 0) {
            var mu = mu
            vDSP_vsadd(start, 1, &mu, start, 1, __count)
        }
    }
#else
    /// Generate random numbers from normal distribution N(mu, sigma).
    /// - Precondition:
    ///   - `count` >= 0
    ///   - `sigma` >= 0
    public func xorshift_normal(start: UnsafeMutablePointer<Float>,
                                count: Int,
                                mu: Float = 0,
                                sigma: Float = 1) {
        precondition(sigma >= 0, "Invalid argument: `sigma` must not be less than 0.")
        precondition(count >= 0, "Invalid argument: `count` must not be less than 0.")
        
        _xorshift_normal(start, Int32(count), mu, sigma)
    }
#endif

/// Sample random number from normal distribution N(mu, sigma).
/// - Precondition:
///   - `sigma` >= 0
public func xorshift_normal(mu: Float = 0, sigma: Float = 1) -> Float {
    precondition(sigma >= 0, "Invalid argument: `sigma` must not be less than 0.")
    
    var ret: Float = 0
    _xorshift_normal(&ret, 1, mu, sigma)
    return ret
}

/// Sample random numbers from normal distribution N(mu, sigma).
///
/// - Note: It's slower than `xorshift_normal` in apple devices, but use less memories.
/// - Precondition:
///   - `count` >= 0
///   - `sigma` >= 0
public func _xorshift_normal(start: UnsafeMutablePointer<Float>,
                             count: Int,
                             mu: Float = 0,
                             sigma: Float = 1) {
    precondition(sigma >= 0, "Invalid argument: `sigma` must not be less than 0.")
    precondition(count >= 0, "Invalid argument: `count` must not be less than 0.")
    
    _xorshift_normal(start, Int32(count), mu, sigma)
}

File: Sources/Cxorswift/xorswift.cpp

// https://en.wikipedia.org/wiki/Xorshift

#include <CoreFoundation/CoreFoundation.h>
#include "xorswift.h"

static uint32_t x = arc4random();
static uint32_t y = arc4random();
static uint32_t z = arc4random();
static uint32_t w = arc4random();

void xorshift(unsigned int *start, int count) {
    uint32_t t1, t2, t3, t4;
    uint32_t *p = start;
    int i;
    
    for(i = 0 ; i < count%4 ; ++i) {
        t1 = x ^ (x << 11);
        x = y; y = z; z = w;
        w = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        *(p++) = w;
    }
    
    for(; i < count ; i += 4) {
        t1 = x ^ (x << 11);
        t2 = y ^ (y << 11);
        t3 = z ^ (z << 11);
        t4 = w ^ (w << 11);
        x = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        y = (x ^ (x >> 19)) ^ (t2 ^ (t2 >> 8));
        z = (y ^ (y >> 19)) ^ (t3 ^ (t3 >> 8));
        w = (z ^ (z >> 19)) ^ (t4 ^ (t4 >> 8));
        *(p++) = x;
        *(p++) = y;
        *(p++) = z;
        *(p++) = w;
    }
}

void xorshift_uniform(float *start, int count, float low, float high) {
    uint32_t t1, t2, t3, t4;
    float *p = start; // in [0, 1)
    float divisor = nextafterf(UINT32_MAX, UINT64_MAX);
    int i;
    
    for(i = 0 ; i < count%4 ; ++i) {
        t1 = x ^ (x << 11);
        x = y; y = z; z = w;
        w = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        *(p++) = (high - low) * (float)w / divisor + low;
    }
    
    for(; i < count ; i += 4) {
        t1 = x ^ (x << 11);
        t2 = y ^ (y << 11);
        t3 = z ^ (z << 11);
        t4 = w ^ (w << 11);
        x = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        y = (x ^ (x >> 19)) ^ (t2 ^ (t2 >> 8));
        z = (y ^ (y >> 19)) ^ (t3 ^ (t3 >> 8));
        w = (z ^ (z >> 19)) ^ (t4 ^ (t4 >> 8));
        *(p++) = (high - low) * (float)x / divisor + low;
        *(p++) = (high - low) * (float)y / divisor + low;
        *(p++) = (high - low) * (float)z / divisor + low;
        *(p++) = (high - low) * (float)w / divisor + low;
    }
}

void _xorshift_normal(float *start, int count, float mu, float sigma) {
    uint32_t t1, t2, t3, t4;
    float *p = start; // ~ N(mu, sigma)
    float x1, x2; // in (0, 1)
    float divisor = nextafterf(UINT32_MAX, UINT64_MAX);
    
    if(count%2 == 1) {
        t1 = x ^ (x << 11);
        x = y; y = z; z = w;
        w = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        x1 = (float)w / divisor + FLT_MIN;
        
        t1 = x ^ (x << 11);
        x = y; y = z; z = w;
        w = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        x2 = (float)w / divisor + FLT_MIN;
        
        *(p++) = sigma * sqrtf(-2*logf(x1)) * cosf(2*M_PI*x2) + mu;
    }
    
    for(int i = 0 ; i < count ; i+=2) {
        t1 = x ^ (x << 11);
        t2 = y ^ (y << 11);
        t3 = z ^ (z << 11);
        t4 = w ^ (w << 11);
        x = (w ^ (w >> 19)) ^ (t1 ^ (t1 >> 8));
        y = (x ^ (x >> 19)) ^ (t2 ^ (t2 >> 8));
        z = (y ^ (y >> 19)) ^ (t3 ^ (t3 >> 8));
        w = (z ^ (z >> 19)) ^ (t4 ^ (t4 >> 8));
        
        x1 = (float)x / divisor + FLT_MIN;
        x2 = (float)y / divisor + FLT_MIN;
        *(p++) = sigma * sqrtf(-2*logf(x1)) * cosf(2*M_PI*x2) + mu;
        
        x1 = (float)z / divisor + FLT_MIN;
        x2 = (float)w / divisor + FLT_MIN;
        *(p++) = sigma * sqrtf(-2*logf(x1)) * cosf(2*M_PI*x2) + mu;
    }
}
